### PR TITLE
mgr/dashboard: Add column filtering feature to cd-table

### PR DESCRIPTION
**Pull request by bk201**
Add column filtering feature to the cd-table component.
- Create filters from columns.
- Create filter options from row data in columns.
- Allow specifying custom functions to do filtering.
- Allow specifying hard-coded options for filters.
- Allow specifying extra filters other than visible columns on table.

Previously, this feature was implemented in inventory component. Refactor
it to move the logic into cd-table, so all tables can use the feature if
needed.

Also, the previous implementation occupied too much space in the table header if there are many filters.

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>
## Usage
- We can enable the filtering feature for a column by enable `filterable` attribute:
```typescript
    this.columns = [
      {
        prop: 'pool_name',
        name: this.i18n('Name'),
        flexGrow: 4,
        cellTransformation: CellTemplate.executing
      },
      {
        prop: 'type',
        name: this.i18n('Type'),
        flexGrow: 2,
        filterable: true // turn on the filtering feature for this column
      },
```
- We can also add some extra filters and custom filter functions: like https://github.com/bk201/ceph/commit/d7f4d0e69063e46017632fbbe2df3b362d25d8b7

## How to test
- Check **Cluster-->Inventory** page by loading some fake data
```
cd /ceph/build
bin/ceph mgr module enable test_orchestrator
bin/ceph orchestrator set backend test_orchestrator
wget https://raw.githubusercontent.com/bk201/ceph-resources/master/issues/40335/dummy.json -O /tmp/dummy.json
bin/ceph test_orchestrator load_data -i /tmp/dummy.json
```
- Check **Cluster-->Configuration** page
## Screenshots
The inventory page is adapted to this change:
inventory

Also the configuration page:
configuration

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html [hunk 1]
```diff
@@ -99,6 +144,27 @@
     </div>
     <!-- end refresh button -->
   </div>
+  <div class="dataTables_header clearfix"
+       *ngIf="toolHeader && columnFiltered">
+    <div class="filter-chips">
+      <span *ngFor="let filter of columnFilters">
+        <span *ngIf="filter.value"
+              class="badge badge-info mr-2 rounded-0">
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html]:
```suggestion
              class="badge badge-info mr-2">
```
angled:
Bildschirmfoto vom 2020-01-09 13-02-19

rounded:
Bildschirmfoto vom 2020-01-09 13-03-05

Bildschirmfoto vom 2020-01-09 13-03-22

Bildschirmfoto vom 2020-01-09 13-04-25

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 2]
```diff
@@ -122,6 +125,10 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
   @Input()
   customCss?: { [css: string]: number | string | ((any) => boolean) };
 
+  // Columns that not displayed but can be used as filters
```
[the remark this answers is not in the document]
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts]: Thanks. Fixed!

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html (hunk 1 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html]: Thanks. Changed!

**Comment by bk201** [comment]
Changes:

- Group two buttons into a Button group, which makes them looks like one:
Screenshot_20200110_151737
- Disable option button when there is no value in a column:
Screenshot_20200110_152027
- Use round corners for filter chips
- Add some padding to the right side of filter chips
Before:
Screenshot_20200110_152820
After:
Screenshot_20200110_152448
- Allow specifying some default applied filters when creating the component. Make sure `basic` level is selected when entering the Configuration page:
Screenshot_20200110_151829

**Comment by votdev** [comment]
One question... Is it possible and does it make sense to be able to deselect a filter by clicking the checked menu in the dropdown again? I know i can click the `x` button in the chip, but when i am currently navigating through the dropdown IMO this would be e great feature.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 3]
```diff
@@ -339,6 +367,112 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
...
+  }
+
+  onSelectFilter(filter: CdTableColumnFilter) {
+    this.selectedFilter = filter;
+  }
+
+  onChangeFilter(filter: CdTableColumnFilter, option?: { raw: string; formatted: string }) {
+    filter.value = option;
+
+    // When using column filters, always clear search field.
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts]: But could you please re-add the custom search after applying the filter?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 4]
```diff
@@ -339,6 +367,112 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
     this.tableColumns = this.columns.filter((c) => !c.isHidden);
   }
 
+  initColumnFilters() {
```
[the remark this answers is not in the document]
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts]: Thanks for reminding. Added.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 5]
```diff
@@ -339,6 +367,112 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
...
+        options: [],
+        value: col.filterInitValue
+          ? this._createColumnFilterOption(col.filterInitValue, col.pipe)
+          : undefined
+      };
+    });
+    this.selectedFilter = _.first(this.columnFilters);
+  }
+
+  _createColumnFilterOption(value: any, pipe?: PipeTransform): { raw: string; formatted: string } {
```
[the remark this answers is not in the document]
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts]: Fixed.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts (hunk 3 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts]: Fixed.

**Comment by bk201** [comment]
Changes:
- Add unit tests for new feature to cd-table.
- Remove an applied filter if select it from the dropdown menu again.
- After clearing filters, if there is text in search field, the results are still based on the search.

**Pull request merged**